core/src/com/unciv/logic/city/CityConstructions.kt

```diff
@@ -413,9 +413,13 @@ class CityConstructions : IsPartOfGameInfoSerialization {
                     }
                 } else if (construction is BaseUnit) {
                     // Production put into upgradable units gets put into upgraded version
-                    if (rejectionReasons.all { it.type == RejectionReasonType.Obsoleted } && construction.upgradesTo != null) {
-                        val upgradedUnitName = city.civ.getEquivalentUnit(construction.upgradesTo!!).name
-                        inProgressConstructions[upgradedUnitName] = (inProgressConstructions[upgradedUnitName] ?: 0) + workDone
+                    val cheapestUpgradeUnit = construction.upgradeUnits(StateForConditionals(city.civ, city))
+                        .mapNotNull { city.civ.gameInfo.ruleset.units[it]?.let { 
```

Comment:
Unifying stuff here isn't exactly easy unless I take the ruleset from... Wait, wtf, since when has the units all had a lateinit ruleset. Not having a centralized ruleset has been the biggest thing plaguing me this entire time

Reply:
Tbh I'm not as aware of how exactly lamda stuff works, but I believe even `(this)->Boolean` would still end up with a `getCheapestUnit { city.civ.getEquivalantUnit(it) }`  which would be less lines of code I guess, but with the added `getRulesetUpgradeUnits`, all that removes is the minByOrNull line which isn't enough of a deal for me to need the helper.